skbio/sequence/_grammared_sequence.py

```diff
@@ -22,11 +22,79 @@
 from ._sequence import Sequence
 
 
-class IUPACSequence(with_metaclass(ABCMeta, Sequence)):
-    """Store biological sequence data conforming to the IUPAC character set.
+class GrammaredSequenceException(TypeError):
+    pass
+
+
+class GrammaredSequenceMeta(ABCMeta, type):
+    def __new__(mcs, name, bases, dct):
+        cls = super(GrammaredSequenceMeta, mcs).__new__(mcs, name, bases, dct)
+
+        # Only perform validation on classes that aren't abstract.
+        if (type(cls.default_gap_char) is not abstractproperty and
```

Comment:
I think this should check that *all* abstract properties are defined, not just `default_gap_char` and `gap_chars`?

Another way to do this would be to validate what you can and skip the validations that can't be performed due to abstractness. This is probably a better way to go because then you can validate what's available, hopefully notifying the user sooner than when the class definition is fully concrete. I'd define some bools at the top and use those as guards to the validations:

```python
default_gap_char_defined = type(cls.default_gap_char) is not abstractproperty
gap_chars_defined = type(cls.gap_chars) is not abstractproperty
degenerate_chars_defined = type(cls.degenerate_chars) is not abstractproperty # this duplicated logic should probably be refactored into a helper method...
...
```